src/reactant.jl

```diff
@@ -0,0 +1,35 @@
+"""
+    Reactor(model)
+
+Container for use with Reactant.jl. Stores the model alongside a compiled forward pass,
+space for the gradient, and a compiled `Enzyme.autodiff` call.
+These compiled functions are created and stored on first use.
+
+Note that unlike `Duplicated(model)`, what is stored is a copy of the model.
+All arrays are copied to Reactant's `ConcreteRArray` type.
+We should make `|> cpu` copy back to ordinary `Array`s, but that doesn't work yet.
+"""
+mutable struct Reactor{M}
+  model::M
+  fwd_compiled
+  fwd_input
+  fwd_count::Int
+  gradient::M
+  grad_compiled
```

Comment:
Longer term it’s actually better to compile the entire update step rather than just the gradient

Reply:
I don't yet see a pretty interface for that. For now I hope that in real use this will be most of the time -- gradients are expensive, and this should capture the entire fwd+back. Optimisers.jl is one or two fused in-place broadcasts per parameter array.

Reply:
Yeah the reason why compiling the entire updating is nice is because we can get rid of intermediate allocations for the gradient/etc and just generate kernels/etc that update the model weights in place

Reply:
Ah ok, that's another step. Then you want the allocation of shadows to also happen within `@compile`, rather than being stored in external Duplicated / Reactor structs.

Reply:
Yeah ideally, because they can get optimized away on the inside of compile